### PR TITLE
REST API for OMP

### PR DESCRIPTION
**Pull request by kaschioudi**

### CODE REVIEW

File: api/v1/series/SerieHandler.inc.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * @ingroup api_v1_user
+ *
+ * @brief Handle API requests for user operations.
+ *
+ */
+
+import('lib.pkp.classes.handler.APIHandler');
+import('classes.core.ServicesContainer');
+
+class SerieHandler extends APIHandler {
```
Comment (asmecher) [inline api/v1/series/SerieHandler.inc.php]: I think this should be SeriesHandler throughout (e.g. `seriesId` rather than `serieId`).

File: api/v1/series/SerieHandler.inc.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
+<?php
+
+/**
+ * @file api/v1/user/SerieHandler.inc.php
```
Comment (asmecher) [inline api/v1/series/SerieHandler.inc.php]: Path needs correction.

File: api/v1/submissions/SubmissionHandler.inc.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+class SubmissionHandler extends PKPSubmissionHandler {
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		$this->_handlerPath = 'submissions';
+		$roles = array(ROLE_ID_MANAGER, ROLE_ID_SUB_EDITOR, ROLE_ID_ASSISTANT, ROLE_ID_REVIEWER, ROLE_ID_AUTHOR);
+		$this->_endpoints = array(
+			'GET' => array (
```
Comment (asmecher) [inline api/v1/submissions/SubmissionHandler.inc.php]: Some of these are identical to those in OJS, right? Is it possible to share more of this in pkp-lib?

File: classes/services/CategoryService.inc.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	 * @param int $contextId
+	 * @param array $args {
+	 * 		@option int parentId
+	 * }
+	 *
+	 * @return array
+	 */
+	public function getCategories($contextId, $args = array()) {
+		$categories = null;
+		$categoryDAO = DAORegistry::getDAO('CategoryDAO');
```
Comment (asmecher) [inline classes/services/CategoryService.inc.php]: Just a convention: it should be `$categoryDao` rather than `$categoryDAO`.

File: docs/dev/swagger.yaml [hunk 5]
```diff
@@ -0,0 +1,1075 @@
...
+        type: string
+
+paths:
+  '/submissions':
+    get:
+      tags:
+        - Submissions
+      summary: Get a list of submissions.
+      description: |
+        **OJS 3.1 Target**:
```
Comment (asmecher) [inline docs/dev/swagger.yaml]: Still refers to OJS. (Is it possible to generalize and move some of this description into `pkp-lib` rather than copying/replacing?)

File: api/v1/submissions/SubmissionHandler.inc.php (hunk 3 above)
Comment (kaschioudi) [reply api/v1/submissions/SubmissionHandler.inc.php]: I have already moved shared handlers' callbacks methods to `PKPSubmissionHandler` class. These are endpoints definitions. Even though some are similar I have left them in subclass for clarity's sake. Do you want me to move common endpoints definitions to the parent class as well?

File: docs/dev/swagger.yaml (hunk 5 above)
Comment (kaschioudi) [reply docs/dev/swagger.yaml]: I have done some research. it's possible to split a swagger file into multiple documents. Where would be a suitable location in `pkp-lib`? Can I store shared definitions in `docs/dev/swagger` folder ?

File: api/v1/submissions/SubmissionHandler.inc.php (hunk 3 above)
Comment (asmecher) [reply api/v1/submissions/SubmissionHandler.inc.php]: It's nothing crucial, so I'll leave it up to you. Where the handler code is all 100% shared in the parent class, I prefer to have the endpoint definition there too.

File: docs/dev/swagger.yaml (hunk 5 above)
Comment (kaschioudi) [reply docs/dev/swagger.yaml]: so I have done the initial work of moving shared `swagger.json` definitions to `pkp-lib`. The problem at the moment is that I have no way to validate the syntax. 

I came across couple of nodejs scripts (https://github.com/mohsen1/multi-file-swagger-example & https://github.com/dzuluaga/json-refs-test) to do so but I was unable to successfully test the swagger document. 

Basically instructions like `$ref: '../../lib/pkp/docs/dev/swagger/produces.yaml'` are supposed to be replaced by the target file content. However when I try it, it's like if the file cannot be resolved.

As per this comment,  `options.relativeBase` needs to be set (https://github.com/whitlockjc/json-refs/issues/78#issuecomment-187930327). It seems that `options.relativeBase` is supported deprecated in general and we should use  `options.location` now (https://github.com/whitlockjc/json-refs/issues/76#issuecomment-301524186)

@NateWr since you are more familiar with NodeJS, any hint ? Thanks.

Comment (NateWr) [reply docs/dev/swagger.yaml]: I've never worked with `json-refs` so I'm not sure. Have you tried the `options.location` approach? No luck? 

When are you trying to validate the syntax? Is this when the tests are run? If you are just looking to validate a JSON file, it might make more sense to lint it? We do that with webpack for the JS in the UI Library, so maybe you could piggy-back on that?

File: docs/dev/swagger.yaml [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    properties:
+      $ref: '../../lib/pkp/docs/dev/swagger/sharedSubmissionProperties.yaml'
+      chapters:
+        type: array
+        items:
+          $ref: '#/definitions/Chapter'
+  SubmissionSummary:
+    type: object
+    properties:
+      $ref: '../../lib/pkp/docs/dev/swagger/sharedSubmissionSummaryProperties.yaml'
```
Comment (NateWr) [inline docs/dev/swagger.yaml]: Let's call these files just `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`. Their position in the `lib/pkp` directory tells us that they are shared definitions.

File: docs/dev/swagger.yaml (hunk 5 above)
Comment (NateWr) [reply docs/dev/swagger.yaml]: I noticed that the online Swagger Editor had a "Convert to JSON" feature. So I dumped your OMP swagger file with it's `$ref`s in, and converted it, and I got refs that look like this:

```
  "definitions": {
    "$ref": "../../lib/pkp/docs/dev/swagger/sharedDefinitions.yaml",
    "Submission": {
      "type": "object",
      "properties": {
        "$ref": "../../lib/pkp/docs/dev/swagger/sharedSubmissionProperties.yaml",
        "chapters": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/Chapter"
          }
        }
      }
    },
```

I know in JSON you have to use `"` not `'`, and maybe because it starts with `$` you need to quote the property too. Could that be the issue you're facing?

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: > Have you tried the options.location approach? No luck?

Yes. But it didn't work.

Alec suggested to move some of the shared swagger spec to pkp-lib. I did some research and saw that it was possible to split a swagger spec into multiple documents via `$ref`. 

After splitting `swagger.yaml` into multiple files, I needed to test the correctness of the spec. We also need to find a way to generate a complete swagger spec (replace refs with their target content) so that a user willing to test from Swagger editor can generate a full swagger spec in a single document. 

Thus instead of having 

> ...
"properties": {
  "$ref": "../../lib/pkp/docs/dev/swagger/sharedSubmissionProperties.yaml",
  "chapters": {
...

You would have 

> ...
"properties": {
  id:
  type: integer
_href:
  type: string
title:
  $ref: '#/definitions/LocaleObject',
  "chapters": {
...


ps: When I move all referenced specs into the same folder or subfolder and run let's say `multi-file-swagger swagger.yaml > index.json` everything works fine, meaning all `$refs` are replaced. But when I move the files to `lib/pkp/docs/dev/swagger`, no luck.

Comment (NateWr) [reply docs/dev/swagger.yaml]: If you have the swagger file in JSON format, you might be able to just use regular node ecosystem tools. This [json-refs](https://www.npmjs.com/package/json-refs) package looks like it has a [command-line tool](https://github.com/whitlockjc/json-refs/blob/master/docs/CLI.md) which might compile a JSON file using it's refs. If that works, we could add it to our build scripts and commit the compiled files to version control.

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: yes that's the plan but I need to figure out the conversion first :sweat:

File: docs/dev/swagger.yaml (hunk 6 above)
Comment (kaschioudi) [reply docs/dev/swagger.yaml]: no problem. I will rename the files.

File: docs/dev/swagger.yaml (hunk 5 above)
Comment (NateWr) [reply docs/dev/swagger.yaml]: Sorry, maybe I'm misunderstanding something, but do you mean converting YAML to JSON? The online [Swagger Editor](http://editor.swagger.io/) can do that from our existing all-in-one YAML file. In the menu go to File > Convert and save as JSON.

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: Sorry, let me try to explain again. 

We had a single `swagger.yaml` containing all the specs. Now we are moving shared specs into other remote files : `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`, etc. 

Let's say a user would like to load OJS REST API swagger specs into Swagger editor. At the moment he could load `swagger.yaml` into the online swagger editor because it's a single file containing all the specs. Once we split into multiple files, `swagger.yaml` will contain references to other files (like `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`) and Swagger editor will not be able to load load these remote spec because these files are located in `pkp-lib` which the online editor do not have access to. 


1. Like you mentioned above, it would be nice to have a `swagger.json` file automatically generated by converting `swagger.yaml`. That's what I am having trouble with. 

2. At the moment I also need to test that the changes I just made (breaking `swagger.yaml` into multiple files) is correct. And to do so I am trying to generate a single file `swagger.yaml` or `swagger.json` from (`swagger.yaml`, `definitions.yaml`, `Submission.yaml` and `SubmissionSummary.yaml`, etc. ). I found a couple of NodeJS tools (see above) that I tried. But instead of replacing `$ref`s with the content of the remote files I still see the paths.

I hope you have a better understanding of the issue now.

Comment (NateWr) [reply docs/dev/swagger.yaml]: Ah, I was thinking that we would abandon YAML and move our specs over to JSON. So I'd see the workflow being:

1. Convert YAML files to JSON files and commit those to version control.
2. When we want a single file from these JSON files, use the CLI tool from the `json-refs` package to generate it based on the JSON.

So instead of trying to use a NodeJS tool to read the YAML files, we just go all-in with JSON. Would that work?

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: Okay with replacing YAML with JSON. but

> When we want a single file from these JSON files, use the CLI tool from the json-refs package to generate it based on the JSON.

That's the issue I am having at the moment. I am unable to create a single file out of these JSON files. see https://github.com/pkp/omp/pull/541#discussion_r191517617

Comment (NateWr) [reply docs/dev/swagger.yaml]: Have you tried the npm package I recommended (`json-refs`)? You mentioned trying two nodejs scripts that were specific to swagger, but I don't think we need anything specific to swagger if we have the JSON files ready.

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: I think I have done so.  I converted `swagger.yaml` to `swagger.json` and then `npm resolve swagger.json`.

Comment (NateWr) [reply docs/dev/swagger.yaml]: Ah ok. Do you have the code with the JSON file somewhere I could try it out?

Comment (NateWr) [reply docs/dev/swagger.yaml]: I think you need to run `json-refs` from the parent directory. So it would be:

```
json-refs food/bar/test.json
```

I found that worked for me when reproducing the error you describe.

(I'm thinking about that disappearing `info` prop. Do `json-refs` actually support property merging? If not, that could be a problem...)

Comment (kaschioudi) [reply docs/dev/swagger.yaml]: it looks like `$ref` must stand alone and should not be used for properties. 
(see https://stackoverflow.com/a/31011359)
